Add EIP: Trustless log and transaction index (#11811)
ethereum/EIPs: EIPS/eip-8304.md

---
eip: 8304
title: Trustless log and transaction index
description: Store the root hashes of index tables in a system contract to allow efficient trustless proofs of log and transaction lookups
author: Zsolt Felföldi (@zsfelfoldi)
discussions-to: https://ethereum-magicians.org/t/eip-8304-trustless-log-and-transaction-index/28824
status: Draft
type: Standards Track
category: Core
created: 2026-06-17

> ---

Reviewer: ```suggestion
requires: 4788
---
```

## Abstract

Store the root hashes of fixed-depth binary tree structures called "index tables" in the storage of a system contract as part of the block processing logic. A new table is generated for each block, then merged into bigger tables indexing multiple blocks. The proposed design prioritizes minimalization of index processing costs during block processing. The EIP has no effect on the bloom filters stored in block headers.

## Motivation

While adding logs has a significantly lower gas cost and is accordingly less resource consuming than writing to the state, currently the protocol defined data structures do not offer even a remotely efficient way of performing content lookups. The original per-block bloom filters were moderately efficient in the beginning and now they are oversaturated and practically useless. This EIP proposes an indexing mechanism that is much cheaper to update than the state tree while allowing content query proofs with a reasonable size, thereby maintaining the low gas cost and increasing the utility of logs.

### Trustless end user access

> Full nodes can maintain their own index structures but most users rely on remote providers that cannot prove the correctness of the results of user queries. Relying on trusted sources for obtaining the output of smart contracts kind of beats the purpose of maintaining those contracts on a distributed trustless ledger. As the processing capacity of Ethereum grows, this information asymmetry just gets worse. Trustless provability throughout the entire stack, from end user applications signing transactions, to the same or other end user applications getting the results they need, should be part of the roadmap.

Reviewer: ```suggestion
Full nodes can maintain their own index structures but most users rely on remote providers that cannot prove the correctness of the results of user queries. Relying on trusted sources for obtaining the output of smart contracts defeats the purpose of maintaining those contracts on a distributed trustless ledger. As the processing capacity of Ethereum grows, this information asymmetry just gets worse. Trustless provability throughout the entire stack, from end user applications signing transactions, to the same or other end user applications getting the results they need, should be part of the roadmap.
```

A bit too informal I think.

00037fcf532c15f0a6db0bd6d0e038bea71d30d808c7d98cb3bf7268a95bf5081b65000000000000002a0000000100000000
0003ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef000000000000002a0000000000000000
0003ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef000000000000002a0000000000000001
0003ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef000000000000002b0000000000000000
00040000000000000000000000004a5e9a3bf35df5e0ea4b4cd3289e0111f1deadf7000000000000002a0000000000000000
00040000000000000000000000004a5e9a3bf35df5e0ea4b4cd3289e0111f1deadf7000000000000002b0000000000000000
0004000000000000000000000000f2a5fd4d5bb24b651d1198bc014941a16f6edde5000000000000002a0000000000000001
0004000000000000000000000000f2a5fd4d5bb24b651d1198bc014941a16f6edde5000000000000002a0000000100000000
00050000000000000000000000004a5e9a3bf35df5e0ea4b4cd3289e0111f1deadf7000000000000002a0000000000000001
0005000000000000000000000000ac8cbe20039c2e9547da9107456179bd4f39a734000000000000002b0000000000000000
0005000000000000000000000000f2a5fd4d5bb24b651d1198bc014941a16f6edde5000000000000002a0000000000000000
```

#### Table root hash calculation

> The table root hash is calculated as the root of the `List[Hash32, entry_count]` SSZ list containing the SHA2 hashes of the binary encoded entries where `entry_count` is the total number of entries in the table.

Reviewer: SHA2 is a bit vague. I assume this is SHA2-256?

Author: Right, fixed.

push1 0x80
shr
push1 0x5c
jumpi
push2 0x0400
dup2
push1 0x04
dup2
div
number
sub
div
dup3
dup3
mul
swap3
push0
calldataload
dup2
dup2
mod
push1 0x5c
jumpi
div
swap1
dup2
sub
not
push2 0x03ff
lt
push1 0x5c
jumpi
mod
add
sload
dup1
iszero
push1 0x5c
jumpi
push0
mstore
push1 0x20
push0
return
jumpdest
push0
push0
revert
jumpdest
push1 0x40
calldataload
push1 0x20
calldataload
push2 0x0400
dup2
dup2
mul
swap2
push0
calldataload
div
mod
add
sstore
stop
```

#### Deployment

A special synthetic address is generated by working backwards from the desired deployment transaction:

```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x3d090",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60758060095f395ff33373fffffffffffffffffffffffffffffffffffffffe1460605760403603605c576020358060801c605c576104008160048104430304828202925f35818106605c5704908103196103ff10605c570601548015605c575f5260205ff35b5f5ffd5b604035602035610400818102915f350406015500",
  "v": "<TBD>",
  "r": "<TBD>",
  "s": "<TBD>",
  "hash": "<TBD>"
}
```

Note, the input in the transaction has a simple constructor prefixing the desired runtime code.

> The sender of the transaction can be calculated as `<TBD>`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals `<TBD>`. This is how `INDEX_CONTRACT_ADDRESS` is determined. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).

Reviewer: ```suggestion
The sender of the transaction can be calculated as <!-- TODO --> `<TBD>`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals <!-- TODO --> `<TBD>`. This is how `INDEX_CONTRACT_ADDRESS` is determined. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).
```

Best to mark TBD items with a comment so the linter can ensure you fix them later on.

## Backwards Compatibility

> This EIP introduces backwards incompatible changes to the block validation rule set. But neither of these changes break anything related to current user activity and experience.

Reviewer: ```suggestion
This EIP introduces backwards incompatible changes to the block validation rule set, but neither of these changes break anything related to current user activity and experience.
```

## Test Cases

> N/A

Reviewer: ```suggestion
```

Just remove the section, but you'll need to add test cases at some point for a Core EIP.
